From 34942d0295e1b5faee2cde6b8f56482f6810f77e Mon Sep 17 00:00:00 2001
From: June Rhodes <jrhodes@redpoint.games>
Date: Thu, 19 Feb 2026 18:42:30 +1100
Subject: [PATCH] Implement checkers for advanced analysis

Includes 'padding' as the first checker, which detects excessive padding due to field ordering.
---
 .../clang/Basic/DiagnosticCommonKinds.td      |   2 +
 clang/lib/Frontend/ClangRulesets.cpp          | 141 +++++++-
 .../Frontend/ClangRulesetsCheckerBase.impl.h  |  42 +++
 .../ClangRulesetsCheckerPadding.impl.h        | 322 ++++++++++++++++++
 4 files changed, 499 insertions(+), 8 deletions(-)
 create mode 100644 clang/lib/Frontend/ClangRulesetsCheckerBase.impl.h
 create mode 100644 clang/lib/Frontend/ClangRulesetsCheckerPadding.impl.h

diff --git a/clang/include/clang/Basic/DiagnosticCommonKinds.td b/clang/include/clang/Basic/DiagnosticCommonKinds.td
index 8bd5ffc219d6..c4f6ebe490a5 100644
--- a/clang/include/clang/Basic/DiagnosticCommonKinds.td
+++ b/clang/include/clang/Basic/DiagnosticCommonKinds.td
@@ -426,6 +426,8 @@ def err_clangrules_ruleset_severity_is_notset : Error<"ruleset '%0' severity mus
 def err_clangrules_rule_name_conflict : Error<"namespaced rule name '%0' is already declared in another .clang-rules file; try using a different namespace to avoid conflicts">;
 def err_clangrules_rule_missing : Error<"namespaced ruleset '%0' requested namespaced rule '%1' but it wasn't known at runtime; make sure it is declared in either the same .clang-rules file, or in a .clang-rules file in a parent directory">;
 def err_clangrules_rule_matcher_parse_failure : Error<"namespaced rule '%0' matcher failed to parse: %1">;
+def err_clangrules_rule_checker_not_found : Error<"namespaced rule '%0' checker is unknown: %1">;
+def err_clangrules_rule_not_one_matcher_or_checker : Error<"namespaced rule '%0' must have exactly one of 'Matcher' and 'Checker'">;
 
 // API notes
 def err_apinotes_message : Error<"%0">;
diff --git a/clang/lib/Frontend/ClangRulesets.cpp b/clang/lib/Frontend/ClangRulesets.cpp
index 7f263c91bd32..3556dbed0562 100644
--- a/clang/lib/Frontend/ClangRulesets.cpp
+++ b/clang/lib/Frontend/ClangRulesets.cpp
@@ -17,6 +17,9 @@
 #include "llvm/Support/Windows/WindowsSupport.h"
 #endif
 
+#include "ClangRulesetsCheckerBase.impl.h"
+#include "ClangRulesetsCheckerPadding.impl.h"
+
 using namespace clang;
 
 #define RULESET_ENABLE_TIMING 0
@@ -238,6 +241,9 @@ struct ClangRulesRule {
   /// The Clang AST matcher to evaluate against top-level declarations in the
   /// AST.
   std::string Matcher;
+  /// The Clang rulesets custom checker that should be used for this rule
+  /// instead of the matcher.
+  std::string Checker;
   /// Specifies the traversal mode for the AST matcher.
   ClangRulesTraversalMode TraversalMode;
   /// The compiler diagnostic message to emit at the callsite when this static
@@ -253,6 +259,9 @@ struct ClangRulesRule {
   /// The runtime matcher value after the matcher expression has been parsed and
   /// loaded by Clang.
   std::optional<clang::ast_matchers::internal::DynTypedMatcher> MatcherParsed;
+  /// The runtime checker value after the checker parameter has been parsed and
+  /// loaded by Clang.
+  std::optional<ClangRulesetsCheckerBase*> CheckerParsed;
   /// If true, this rule only applies when the compilation triple targets
   /// Windows. This can be used for rules which match on Windows-specific AST
   /// nodes (such as '__declspec(dllexport)').
@@ -400,11 +409,12 @@ struct ScalarEnumerationTraits<
 template <> struct MappingTraits<clang::rulesets::config::ClangRulesRule> {
   static void mapping(IO &IO, clang::rulesets::config::ClangRulesRule &Rule) {
     IO.mapRequired("Name", Rule.Name);
-    IO.mapRequired("Matcher", Rule.Matcher);
+    IO.mapOptional("Matcher", Rule.Matcher);
+    IO.mapOptional("Checker", Rule.Checker);
     IO.mapOptional("TraversalMode", Rule.TraversalMode,
                    clang::rulesets::config::ClangRulesTraversalMode::CRTM_AsIs);
-    IO.mapRequired("ErrorMessage", Rule.ErrorMessage);
-    IO.mapRequired("Callsite", Rule.Callsite);
+    IO.mapOptional("ErrorMessage", Rule.ErrorMessage);
+    IO.mapOptional("Callsite", Rule.Callsite);
     IO.mapOptional("Hints", Rule.Hints);
     IO.mapOptional("WindowsOnly", Rule.WindowsOnly, false);
     IO.mapOptional("Debug", Rule.Debug, false);
@@ -477,6 +487,24 @@ template <> struct MappingTraits<clang::rulesets::config::ClangRules> {
 
 namespace clang::rulesets {
 
+template<typename T>
+static std::pair<std::string, std::unique_ptr<ClangRulesetsCheckerBase>> createCheckerDefinition() {
+  
+}
+
+static std::map<std::string, std::unique_ptr<ClangRulesetsCheckerBase>> CheckerDefinitions;
+
+template<typename T>
+class CheckerDefinitionRegister {
+public:
+  CheckerDefinitionRegister() {
+    T* Instance = new T();
+    CheckerDefinitions[Instance->getName()] = std::unique_ptr<ClangRulesetsCheckerBase>(Instance);
+  }
+};
+
+static CheckerDefinitionRegister<ClangRulesetsCheckerPadding> RegisterPadding;
+
 struct ClangRulesetsEffectiveRule {
   // Pointer to memory inside a loaded config::ClangRules.
   config::ClangRulesRule *Rule;
@@ -1271,6 +1299,7 @@ private:
         }
 
         // Attempt to parse the matcher expression.
+        if (!Rule.Matcher.empty())
         {
           clang::ast_matchers::dynamic::Diagnostics ParseDiag;
           llvm::StringRef MatcherRef(Rule.Matcher);
@@ -1295,6 +1324,31 @@ private:
             // MatcherParsed in this case.
           }
         }
+
+        // Attempt to parse the checker value.
+        if (!Rule.Checker.empty()) {
+          if (auto FoundChecker = CheckerDefinitions.find(Rule.Checker); FoundChecker != CheckerDefinitions.end()) {
+            Rule.CheckerParsed = FoundChecker->second.get();
+          } else {
+            SrcMgr.getDiagnostics().Report(
+                SrcMgr.getLocForStartOfFile(FileID),
+                diag::err_clangrules_rule_checker_not_found)
+                << NamespacedName << Rule.Checker;
+            StillValid = false;
+            continue;
+          }
+        }
+
+        // Make sure we either have a matcher or checker.
+        if ((!Rule.MatcherParsed.has_value() && !Rule.CheckerParsed.has_value()) ||
+            (Rule.MatcherParsed.has_value() && Rule.CheckerParsed.has_value())) {
+          SrcMgr.getDiagnostics().Report(
+              SrcMgr.getLocForStartOfFile(FileID),
+              diag::err_clangrules_rule_not_one_matcher_or_checker)
+              << NamespacedName;
+          StillValid = false;
+          continue;
+        }
       }
 
       // Go through rulesets and namespace both their name and any unprefixed
@@ -1546,11 +1600,20 @@ private:
       // and generate diagnostic IDs so that code can use pragmas to control
       // them.
       for (const auto &EffectiveRule : EffectiveConfig->EffectiveRules) {
-        this->CI.getDiagnostics().getDiagnosticIDs()->getCustomDiagID(
-            convertDiagnosticLevel(EffectiveRule.second.Severity),
-            (EffectiveRule.second.Rule->ErrorMessage + " [-W" +
-             EffectiveRule.second.Rule->Name + "]"),
-            EffectiveRule.second.Rule->Name);
+        if (EffectiveRule.second.Rule->Checker.empty()) {
+          this->CI.getDiagnostics().getDiagnosticIDs()->getCustomDiagID(
+              convertDiagnosticLevel(EffectiveRule.second.Severity),
+              (EffectiveRule.second.Rule->ErrorMessage + " [-W" +
+              EffectiveRule.second.Rule->Name + "]"),
+              EffectiveRule.second.Rule->Name);
+        } else {
+          // Checkers always take the message as an argument to the diagnostic via %0.
+          this->CI.getDiagnostics().getDiagnosticIDs()->getCustomDiagID(
+              convertDiagnosticLevel(EffectiveRule.second.Severity),
+              ("%0 [-W" +
+              EffectiveRule.second.Rule->Name + "]"),
+              EffectiveRule.second.Rule->Name);
+        }
       }
 
       // Otherwise, this is the effective config for this directory.
@@ -1585,6 +1648,7 @@ private:
 
   class InstantiatedMatcher {
   private:
+
     class InstantiatedMatcherCallback
         : public clang::ast_matchers::MatchFinder::MatchCallback {
     private:
@@ -1672,6 +1736,58 @@ private:
       }
     };
 
+    class InstantiatedCheckerCallback
+        : public clang::ast_matchers::MatchFinder::MatchCallback {
+    private:
+      llvm::sys::SmartMutex<true> &Mutex;
+      ASTContext &AST;
+      const ClangRulesetsEffectiveRule &EffectiveRule;
+
+    public:
+      InstantiatedCheckerCallback(
+          llvm::sys::SmartMutex<true> &InMutex, ASTContext &InAST,
+          const ClangRulesetsEffectiveRule &InEffectiveRule)
+          : Mutex(InMutex), AST(InAST), EffectiveRule(InEffectiveRule){};
+
+      virtual void run(const clang::ast_matchers::MatchFinder::MatchResult
+                           &Result) override {
+        this->EffectiveRule.Rule->CheckerParsed.value()->processMatchResult(
+          this->AST,
+          Result,
+          [this](const std::vector<ClangRulesetsCheckerDiagnostic>& Diagnostics) {
+            if (Diagnostics.size() == 0) {
+              return;
+            }
+
+            // Obtain lock.
+            this->Mutex.lock();
+
+            // Emit reported diagnostics.
+            for (const auto& Diagnostic : Diagnostics) {
+              if (!Diagnostic.IsNote) {
+                clang::DiagnosticIDs::Level DiagnosticLevel =
+                    convertDiagnosticLevel(this->EffectiveRule.Severity);
+                auto CallsiteDiagID =
+                    this->AST.getDiagnostics().getDiagnosticIDs()->getExistingCustomDiagID(
+                            this->EffectiveRule.Rule->Name, DiagnosticLevel);
+                assert(
+                    CallsiteDiagID
+                        .has_value() /* expected diagnostics to have been created */);
+                this->AST.getDiagnostics().Report(Diagnostic.Location, CallsiteDiagID.value()) << Diagnostic.Message;
+              } else {
+                auto HintDiagID =
+                    this->AST.getDiagnostics().getDiagnosticIDs()->getCustomDiagID(
+                        clang::DiagnosticIDs::Note, Diagnostic.Message);
+                this->AST.getDiagnostics().Report(Diagnostic.Location, HintDiagID);
+              }
+            }
+
+            // Release lock.
+            this->Mutex.unlock();
+          });
+      }
+    };
+
     std::unique_ptr<ast_matchers::MatchFinder> Finder;
     llvm::DenseMap<const ClangRulesetsEffectiveRule *,
                    clang::ast_matchers::MatchFinder::MatchCallback *>
@@ -1703,6 +1819,15 @@ private:
         this->Finder->addDynamicMatcher(*Rule->MatcherParsed, Callback);
         this->Callbacks[&EffectiveRule] = Callback;
       }
+      else if (Rule->CheckerParsed.has_value()) {
+        auto *Callback = new InstantiatedCheckerCallback(this->Mutex, this->AST,
+                                                         EffectiveRule);
+        RULESET_TRACE_RULESET("Adding dynamic checker to finder.\n");
+        Rule->CheckerParsed.value()->registerMatcherWithFinder(
+          *this->Finder,
+          Callback);
+        this->Callbacks[&EffectiveRule] = Callback;
+      }
     }
 
     void match(const OptionalFileEntryRef &FileEntry, clang::Decl *Decl) {
diff --git a/clang/lib/Frontend/ClangRulesetsCheckerBase.impl.h b/clang/lib/Frontend/ClangRulesetsCheckerBase.impl.h
new file mode 100644
index 000000000000..808e91a9d414
--- /dev/null
+++ b/clang/lib/Frontend/ClangRulesetsCheckerBase.impl.h
@@ -0,0 +1,42 @@
+#pragma once
+
+#include "clang/AST/ASTConsumer.h"
+#include "clang/Frontend/CompilerInstance.h"
+
+namespace clang::rulesets {
+
+struct ClangRulesetsCheckerDiagnostic {
+public:
+  ClangRulesetsCheckerDiagnostic(
+    const clang::SourceLocation& InLocation,
+    bool InIsNote,
+    llvm::StringRef InMessage)
+    : Location(InLocation)
+    , IsNote(InIsNote)
+    , Message(InMessage)
+  {
+  }
+
+  clang::SourceLocation Location;
+  bool IsNote;
+  llvm::StringRef Message;
+};
+
+using ClangRulesetsCheckerReportDiagnostics = std::function<void(
+      const std::vector<ClangRulesetsCheckerDiagnostic>& Diagnostics)>;
+
+class ClangRulesetsCheckerBase {
+public:
+  virtual ~ClangRulesetsCheckerBase() = default;
+
+  virtual std::string getName() const = 0;
+
+  virtual void registerMatcherWithFinder(clang::ast_matchers::MatchFinder& Finder, clang::ast_matchers::MatchFinder::MatchCallback *Action) const = 0;
+
+  virtual void processMatchResult(
+    ASTContext &AST, 
+    const clang::ast_matchers::MatchFinder::MatchResult &Result,
+    const ClangRulesetsCheckerReportDiagnostics& ReportDiagnostics) const = 0;
+};
+
+} // namespace clang::rulesets
diff --git a/clang/lib/Frontend/ClangRulesetsCheckerPadding.impl.h b/clang/lib/Frontend/ClangRulesetsCheckerPadding.impl.h
new file mode 100644
index 000000000000..724c78814281
--- /dev/null
+++ b/clang/lib/Frontend/ClangRulesetsCheckerPadding.impl.h
@@ -0,0 +1,322 @@
+#pragma once
+
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecordLayout.h"
+#include "clang/Frontend/CompilerInstance.h"
+
+namespace clang::rulesets {
+
+class ClangRulesetsCheckerPadding : public ClangRulesetsCheckerBase {
+public:
+  virtual ~ClangRulesetsCheckerPadding() override = default;
+
+  virtual std::string getName() const override {
+    return "padding";
+  }
+
+  virtual void registerMatcherWithFinder(clang::ast_matchers::MatchFinder& Finder, clang::ast_matchers::MatchFinder::MatchCallback *Action) const override {
+    using namespace clang::ast_matchers;
+
+    Finder.addMatcher(recordDecl().bind("record"), Action);
+    Finder.addMatcher(varDecl().bind("var"), Action);
+  }
+
+  virtual void processMatchResult(
+    ASTContext &AST, 
+    const clang::ast_matchers::MatchFinder::MatchResult &Result,
+    const ClangRulesetsCheckerReportDiagnostics& ReportDiagnostics) const override {
+    if (auto* RD = Result.Nodes.getNodeAs<RecordDecl>("record")) {
+      this->visitRecord(AST, ReportDiagnostics, RD);
+    }
+    if (auto* VD = Result.Nodes.getNodeAs<VarDecl>("var")) {
+      this->visitVariable(AST, ReportDiagnostics, VD);
+    }
+  }
+
+private:
+  static constexpr int32_t AllowedPad = 0;
+
+  /// Look for records of overly padded types. If padding *
+  /// PadMultiplier exceeds AllowedPad, then generate a report.
+  /// PadMultiplier is used to share code with the array padding
+  /// checker.
+  void visitRecord(ASTContext &AST, const ClangRulesetsCheckerReportDiagnostics& ReportDiagnostics, const RecordDecl *RD, uint64_t PadMultiplier = 1) const {
+    if (shouldSkipDecl(AST, RD))
+      return;
+
+    // TODO: Figure out why we are going through declarations and not only
+    // definitions.
+    if (!(RD = RD->getDefinition()))
+      return;
+
+    // This is the simplest correct case: a class with no fields and one base
+    // class. Other cases are more complicated because of how the base classes
+    // & fields might interact, so we don't bother dealing with them.
+    // TODO: Support other combinations of base classes and fields.
+    if (auto *CXXRD = dyn_cast<CXXRecordDecl>(RD))
+      if (CXXRD->field_empty() && CXXRD->getNumBases() == 1)
+        return visitRecord(AST, ReportDiagnostics, CXXRD->bases().begin()->getType()->getAsRecordDecl(),
+                           PadMultiplier);
+
+    auto &ASTContext = RD->getASTContext();
+    const ASTRecordLayout &RL = ASTContext.getASTRecordLayout(RD);
+    assert(llvm::isPowerOf2_64(RL.getAlignment().getQuantity()));
+
+    CharUnits BaselinePad = calculateBaselinePad(RD, ASTContext, RL);
+    if (BaselinePad.isZero())
+      return;
+
+    CharUnits OptimalPad;
+    SmallVector<const FieldDecl *, 20> OptimalFieldsOrder;
+    std::tie(OptimalPad, OptimalFieldsOrder) =
+        calculateOptimalPad(RD, ASTContext, RL);
+
+    CharUnits DiffPad = PadMultiplier * (BaselinePad - OptimalPad);
+    if (DiffPad.getQuantity() <= AllowedPad) {
+      assert(!DiffPad.isNegative() && "DiffPad should not be negative");
+      // There is not enough excess padding to trigger a warning.
+      return;
+    }
+    reportRecord(AST, ReportDiagnostics, RD, BaselinePad, OptimalPad, OptimalFieldsOrder);
+  }
+
+  /// Look for arrays of overly padded types. If the padding of the
+  /// array type exceeds AllowedPad, then generate a report.
+  void visitVariable(ASTContext &AST, const ClangRulesetsCheckerReportDiagnostics& ReportDiagnostics, const VarDecl *VD) const {
+    const ArrayType *ArrTy = VD->getType()->getAsArrayTypeUnsafe();
+    if (ArrTy == nullptr)
+      return;
+    uint64_t Elts = 0;
+    if (const ConstantArrayType *CArrTy = dyn_cast<ConstantArrayType>(ArrTy))
+      Elts = CArrTy->getZExtSize();
+    if (Elts == 0)
+      return;
+    const RecordType *RT = ArrTy->getElementType()->getAs<RecordType>();
+    if (RT == nullptr)
+      return;
+
+    // TODO: Recurse into the fields to see if they have excess padding.
+    visitRecord(AST, ReportDiagnostics, RT->getDecl(), Elts);
+  }
+
+  bool shouldSkipDecl(ASTContext &AST, const RecordDecl *RD) const {
+    // TODO: Figure out why we are going through declarations and not only
+    // definitions.
+    if (!(RD = RD->getDefinition()))
+      return true;
+    auto Location = RD->getLocation();
+    // If the construct doesn't have a source file, then it's not something
+    // we want to diagnose.
+    if (!Location.isValid())
+      return true;
+    SrcMgr::CharacteristicKind Kind =
+        AST.getSourceManager().getFileCharacteristic(Location);
+    // Throw out all records that come from system headers.
+    if (Kind != SrcMgr::C_User)
+      return true;
+
+    // Not going to attempt to optimize unions.
+    if (RD->isUnion())
+      return true;
+    if (auto *CXXRD = dyn_cast<CXXRecordDecl>(RD)) {
+      // Tail padding with base classes ends up being very complicated.
+      // We will skip objects with base classes for now, unless they do not
+      // have fields.
+      // TODO: Handle more base class scenarios.
+      if (!CXXRD->field_empty() && CXXRD->getNumBases() != 0)
+        return true;
+      if (CXXRD->field_empty() && CXXRD->getNumBases() != 1)
+        return true;
+      // Virtual bases are complicated, skipping those for now.
+      if (CXXRD->getNumVBases() != 0)
+        return true;
+      // Can't layout a template, so skip it. We do still layout the
+      // instantiations though.
+      if (CXXRD->getTypeForDecl()->isDependentType())
+        return true;
+      if (CXXRD->getTypeForDecl()->isInstantiationDependentType())
+        return true;
+    }
+    // How do you reorder fields if you haven't got any?
+    else if (RD->field_empty())
+      return true;
+
+    auto IsTrickyField = [](const FieldDecl *FD) -> bool {
+      // Bitfield layout is hard.
+      if (FD->isBitField())
+        return true;
+
+      // Variable length arrays are tricky too.
+      QualType Ty = FD->getType();
+      if (Ty->isIncompleteArrayType())
+        return true;
+      return false;
+    };
+
+    if (llvm::any_of(RD->fields(), IsTrickyField))
+      return true;
+    return false;
+  }
+
+  static CharUnits calculateBaselinePad(const RecordDecl *RD,
+                                        const ASTContext &ASTContext,
+                                        const ASTRecordLayout &RL) {
+    CharUnits PaddingSum;
+    CharUnits Offset = ASTContext.toCharUnitsFromBits(RL.getFieldOffset(0));
+    for (const FieldDecl *FD : RD->fields()) {
+      // Skip field that is a subobject of zero size, marked with
+      // [[no_unique_address]] or an empty bitfield, because its address can be
+      // set the same as the other fields addresses.
+      if (FD->isZeroSize(ASTContext))
+        continue;
+      // This checker only cares about the padded size of the
+      // field, and not the data size. If the field is a record
+      // with tail padding, then we won't put that number in our
+      // total because reordering fields won't fix that problem.
+      CharUnits FieldSize = ASTContext.getTypeSizeInChars(FD->getType());
+      auto FieldOffsetBits = RL.getFieldOffset(FD->getFieldIndex());
+      CharUnits FieldOffset = ASTContext.toCharUnitsFromBits(FieldOffsetBits);
+      PaddingSum += (FieldOffset - Offset);
+      Offset = FieldOffset + FieldSize;
+    }
+    PaddingSum += RL.getSize() - Offset;
+    return PaddingSum;
+  }
+
+  /// Optimal padding overview:
+  /// 1.  Find a close approximation to where we can place our first field.
+  ///     This will usually be at offset 0.
+  /// 2.  Try to find the best field that can legally be placed at the current
+  ///     offset.
+  ///   a.  "Best" is the largest alignment that is legal, but smallest size.
+  ///       This is to account for overly aligned types.
+  /// 3.  If no fields can fit, pad by rounding the current offset up to the
+  ///     smallest alignment requirement of our fields. Measure and track the
+  //      amount of padding added. Go back to 2.
+  /// 4.  Increment the current offset by the size of the chosen field.
+  /// 5.  Remove the chosen field from the set of future possibilities.
+  /// 6.  Go back to 2 if there are still unplaced fields.
+  /// 7.  Add tail padding by rounding the current offset up to the structure
+  ///     alignment. Track the amount of padding added.
+
+  static std::pair<CharUnits, SmallVector<const FieldDecl *, 20>>
+  calculateOptimalPad(const RecordDecl *RD, const ASTContext &ASTContext,
+                      const ASTRecordLayout &RL) {
+    struct FieldInfo {
+      CharUnits Align;
+      CharUnits Size;
+      const FieldDecl *Field;
+      bool operator<(const FieldInfo &RHS) const {
+        // Order from small alignments to large alignments,
+        // then large sizes to small sizes.
+        // then large field indices to small field indices
+        return std::make_tuple(Align, -Size,
+                               Field ? -static_cast<int>(Field->getFieldIndex())
+                                     : 0) <
+               std::make_tuple(
+                   RHS.Align, -RHS.Size,
+                   RHS.Field ? -static_cast<int>(RHS.Field->getFieldIndex())
+                             : 0);
+      }
+    };
+    SmallVector<FieldInfo, 20> Fields;
+    auto GatherSizesAndAlignments = [](const FieldDecl *FD) {
+      FieldInfo RetVal;
+      RetVal.Field = FD;
+      auto &Ctx = FD->getASTContext();
+      auto Info = Ctx.getTypeInfoInChars(FD->getType());
+      RetVal.Size = FD->isZeroSize(Ctx) ? CharUnits::Zero() : Info.Width;
+      RetVal.Align = Info.Align;
+      assert(llvm::isPowerOf2_64(RetVal.Align.getQuantity()));
+      if (auto Max = FD->getMaxAlignment())
+        RetVal.Align = std::max(Ctx.toCharUnitsFromBits(Max), RetVal.Align);
+      return RetVal;
+    };
+    std::transform(RD->field_begin(), RD->field_end(),
+                   std::back_inserter(Fields), GatherSizesAndAlignments);
+    llvm::sort(Fields);
+    // This lets us skip over vptrs and non-virtual bases,
+    // so that we can just worry about the fields in our object.
+    // Note that this does cause us to miss some cases where we
+    // could pack more bytes in to a base class's tail padding.
+    CharUnits NewOffset = ASTContext.toCharUnitsFromBits(RL.getFieldOffset(0));
+    CharUnits NewPad;
+    SmallVector<const FieldDecl *, 20> OptimalFieldsOrder;
+    while (!Fields.empty()) {
+      unsigned TrailingZeros =
+          llvm::countr_zero((unsigned long long)NewOffset.getQuantity());
+      // If NewOffset is zero, then countTrailingZeros will be 64. Shifting
+      // 64 will overflow our unsigned long long. Shifting 63 will turn
+      // our long long (and CharUnits internal type) negative. So shift 62.
+      long long CurAlignmentBits = 1ull << (std::min)(TrailingZeros, 62u);
+      CharUnits CurAlignment = CharUnits::fromQuantity(CurAlignmentBits);
+      FieldInfo InsertPoint = {CurAlignment, CharUnits::Zero(), nullptr};
+
+      // In the typical case, this will find the last element
+      // of the vector. We won't find a middle element unless
+      // we started on a poorly aligned address or have an overly
+      // aligned field.
+      auto Iter = llvm::upper_bound(Fields, InsertPoint);
+      if (Iter != Fields.begin()) {
+        // We found a field that we can layout with the current alignment.
+        --Iter;
+        NewOffset += Iter->Size;
+        OptimalFieldsOrder.push_back(Iter->Field);
+        Fields.erase(Iter);
+      } else {
+        // We are poorly aligned, and we need to pad in order to layout another
+        // field. Round up to at least the smallest field alignment that we
+        // currently have.
+        CharUnits NextOffset = NewOffset.alignTo(Fields[0].Align);
+        NewPad += NextOffset - NewOffset;
+        NewOffset = NextOffset;
+      }
+    }
+    // Calculate tail padding.
+    CharUnits NewSize = NewOffset.alignTo(RL.getAlignment());
+    NewPad += NewSize - NewOffset;
+    return {NewPad, std::move(OptimalFieldsOrder)};
+  }
+
+  void reportRecord(
+      ASTContext &AST, const ClangRulesetsCheckerReportDiagnostics& ReportDiagnostics, const RecordDecl *RD, CharUnits BaselinePad, CharUnits OptimalPad,
+      const SmallVector<const FieldDecl *, 20> &OptimalFieldsOrder) const {
+    SmallString<100> Buf;
+    llvm::raw_svector_ostream Os(Buf);
+    Os << "excessive padding in '";
+    Os << QualType::getAsString(RD->getTypeForDecl(), Qualifiers(),
+                                LangOptions())
+       << "'";
+
+    if (auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
+      // TODO: make this show up better in the console output and in
+      // the HTML. Maybe just make it show up in HTML like the path
+      // diagnostics show.
+      SourceLocation ILoc = TSD->getPointOfInstantiation();
+      if (ILoc.isValid())
+        Os << " instantiated here: "
+           << ILoc.printToString(AST.getSourceManager());
+    }
+
+    Os << " (" << BaselinePad.getQuantity() << " padding bytes, where "
+       << OptimalPad.getQuantity() << " is optimal). "
+       << "optimal fields order: ";
+    bool First = true;
+    for (const auto *FD : OptimalFieldsOrder) {
+      if (!First) {
+        Os << ", ";
+      } else {
+        First = false;
+      }
+      Os << FD->getName();
+    }
+    Os << ". consider reordering the fields or adding explicit padding "
+          "members.";
+
+    ReportDiagnostics(std::vector<ClangRulesetsCheckerDiagnostic> {
+      ClangRulesetsCheckerDiagnostic(RD->getSourceRange().getBegin(), false, Os.str())
+    });
+  }
+};
+
+} // namespace clang::rulesets
-- 
2.50.1.windows.1

